### PR TITLE
[REVIEW] Allow dask_cudf.read_parquet to handle partitioned data

### PR DESCRIPTION
**Pull request by rjzamora**
This is a *short-term* fix for the `dask_cudf.read_parquet` bug described in [#2829](https://github.com/rapidsai/cudf/issues/2829) (the solution can be revisited once #2845  is addressed).

These changes allow dask_cudf to handle partitioned datasets using the available `piece.partition_keys` and `partitions` objects from pyarrow. The logic is mostly copied [from `apache/arrow/python/pyarrow/parquet.py`](https://github.com/apache/arrow/blob/545fb348f4d4e33ef52e9af551e2b7cda241b31c/python/pyarrow/parquet.py#L593).

**Note**:  `dask_cudf.read_parquet(...).compute()` does **not** return categorical columns for the partitions (while `dask.dataframe` **does**).  I may be able to resolve this discrepancy soon, but so far I have been running into cudf `merge` errors.  Due to this discrepancy, `test_roundtrip_from_dask_partitioned` is a bit more convoluted than I would like.

### CODE REVIEW

File: python/dask_cudf/dask_cudf/io/parquet.py [hunk 1]
```diff
@@ -55,6 +55,22 @@ def read_partition(
...
+        if len(piece.partition_keys) > 0:
+            # Partitioning logic copied from
+            # (apache/arrow/python/pyarrow/parquet.py)
+            if partitions is None:
+                raise ValueError("Must pass partition sets")
+            import numpy as np
+            import pyarrow as pa
+
+            for i, (name, index2) in enumerate(piece.partition_keys):
+                indices = np.array([index2], dtype="i4").repeat(len(df))
```
Comment (kkraus14) [inline python/dask_cudf/dask_cudf/io/parquet.py]: This seems expensive to do for a large dataframe, should we do this on the gpu instead?

File: python/dask_cudf/dask_cudf/io/parquet.py [hunk 2]
```diff
@@ -55,6 +55,22 @@ def read_partition(
...
+                dictionary = partitions.levels[i].dictionary
+                arr = pa.DictionaryArray.from_arrays(
+                    indices, dictionary
+                ).to_pandas()
```
Comment (kkraus14) [inline python/dask_cudf/dask_cudf/io/parquet.py]: Similarly here, should we construct a categorical column directly on the GPU instead?

File: python/dask_cudf/dask_cudf/io/parquet.py (hunk 1 above)
Comment (rjzamora) [reply python/dask_cudf/dask_cudf/io/parquet.py]: Yes - Thanks for taking a look!  I definitley plan to improve this today.

**Comment by rjzamora** [comment]
**Update**:  I am having a bit of trouble modifying the `read_partition` code to construct categorical columns for the partitions (to make behavior directly comparable with main-line Dask). I have tried several different approaches to construct a `CategoricalColumn` - For example:

```python
if len(piece.partition_keys) > 0:
    if partitions is None:
        raise ValueError("Must pass partition sets")
    for i, (name, index2) in enumerate(piece.partition_keys):
        categories = [val.as_py() for val in partitions.levels[i].dictionary]
        sr = cudf.Series(index2).astype(type(index2)).repeat(len(df))
        df[name] = CategoricalColumn(
            data=sr._column.data,
            categories=categories,
            ordered=False,
        )
```

The above aproach seems to fail 50% of the time at random.  The occational failure being that the final columns sometimes revert to the integer (code) values rather than categoricals.  

```
a =        month  day  data  year
index                        
0          1    1     0  2018
1          2    1     0  2019
2          3    1     0  2019
3          3    2     0  2019
4          3    2     0  2020
5          2    1     0  2021
b =        month  day  data  year
index                        
0          1    1     0     0
1          2    1     0     1
2          3    1     0     1
3          3    2     0     1
4          3    2     0     2
5          2    1     0     3
```

Note that the above approach works every time when the written dataset includes one partition (e.g. using `ddf2 = dd.from_pandas(df, npartitions=2)` in `test_roundtrip_from_dask_partitioned`)...  I will revisit this problem later/tomorrow (any intuition/suggestions are welcome).

**Comment by rjzamora** [comment]
If one of the cudf/python gurus has a free moment to take a look (i.e. @kkraus14, @shwina):  Does it look like I am building a `CategoricalColumn` correctly in [the above comment/snippet](https://github.com/rapidsai/cudf/pull/2850#issuecomment-534263469)?

**Comment by shwina** [comment]
Looks right and the failure you're seeing is definitely strange. Do you have a reproducer I can tinker with?

**Comment by rjzamora** [comment]
> Looks right and the failure you're seeing is definitely strange. Do you have a reproducer I can tinker with?

Thanks @shwina!  In that case, I will definitely try to come up with a reproducer and get back to you :)

**Comment by rjzamora** [comment]
There are currently known CI problems here, but I confirmed locally that #2869 *should* resolve the previously-discussed categorical bug (thanks @shwina!)

**Comment by kkraus14** [comment]
rerun tests

**Pull request merged**